pyfr/plugins/acoustics.py

```diff
@@ -0,0 +1,785 @@
+        ltsub = self.cfg.getfloat(cfgsect, 'window-length', ltsub_def)
+        shift = self.cfg.getfloat(cfgsect, 'time-shift', 0.5)
+        window_func = self.cfg.get(cfgsect, 'window', 'hanning')
+        psd_scale_mode = self.cfg.get(cfgsect,'psd-scaling-mode',
+                                            'density')
+        samplstps = self.cfg.getint(cfgsect, 'nsamplsteps', 0)
+        timeparam = TimeParam(self.ttol, dtsub, ltsub, shift,
+                                   samplstps, psd_scale_mode,
+                                   window_func)
+
+        obsv_pts = np.array(self.cfg.getliteral(self.cfgsect,
+                                                     'observer-locs'))
+
+        uinf = defaultdict(list)
+        for pvvar in self.privarmap:
+            if pvvar in ['u', 'v', 'w']:
+                uinf['u'].append(npeval(self.cfg.getexpr(cfgsect, pvvar),
+                                        self.constvars))
+            else:
+                uinf[pvvar] = npeval(self.cfg.getexpr(cfgsect, pvvar),
+                                     self.constvars)
+
+        if self._artificial_compress:
+            uinf['c'] = np.sqrt(self.constvars['ac-zeta'])
+            uinf['rho'] = self.uinf['p']/self.constvars['ac-zeta']
+        else:
+            gmma = self.constvars['gamma']
+            uinf['c'] = np.sqrt(gmma*uinf['p']/uinf['rho'])
+        uinf['Mach'] = uinf['u']/uinf['c']
+
+        self.fwhsolver = FwhSolver(timeparam, obsv_pts, *surf, uinf,
+                                   self.privarmap, self.surftype
+                                  )
+
+    def _sample_surf(self, mesh, rallocs, ele_map):
+        periodic = self.cfg.getbool(self.cfgsect, 'inclperiodic', False)
+
+        expr = self.cfg.get(self.cfgsect, 'endcaps', None)
+        endcaps = []
+
+        # Factor out the individual endcap region expressions
+        from ast import literal_eval
+        expr = re.sub(
+            r'(\w+)\((' + match_paired_paren('()') + r')\)',
+            lambda m: endcaps.append(literal_eval(m.groups()[1])) or
+            f'e{len(endcaps) - 1}',
+            expr
+        )
+
+        # Initialise by assuming all faces in region are on the surface
+        sfaces = set()
+        for k, eidls in self._ele_region_data.items():
+            etype  = k.removesuffix('_idxs')
+            nfaces = len(subclass_where(BaseShape, name=etype).faces)
+            sfaces.update((etype, i, j) for i in eidls for j in range(nfaces))
+
+        # Eliminate nndcap faces if any
+        endcap_sfaces = self._sfaces_endcaps(ele_map, sfaces, endcaps)
+        sfaces.difference_update(endcap_sfaces)
+
+        # Loop over interface, boundary and mpi faces and find those on surface
+        surfeset = set()
+        self._eidxs = defaultdict(list)
+
+        e, self._int_eidxs = self._int_faces(mesh, rallocs, surfeset, sfaces,
+                                             periodic, self.bcs)
+        dd_list_merge(self._eidxs, e)
```

Comment:
I think this can be simplified, _eidxs doesn't get used anymore in the generation of the various faces and is now only a union of _int_eidxs, _mpi_eidx, _bnd_eidxs

Reply:
> @mhawwary I have moved the region stuff over now and changed the MPI. I think it's best to keep working on this version as I've done significant refactoring already. Let me know if there are any other modifications though that you think would be appropriate.

That seems good. It looks like you started the refactoring on a version that already have most of what is in that private repo. I did not notice that in the beginning. I will let you know if I have any comments but looks good so far. Thanks for working on this.